CRM/Contract/Form/Modify.php

```diff
@@ -11,6 +11,13 @@ class CRM_Contract_Form_Modify extends CRM_Core_Form{
 
   function preProcess(){
 
+    // If we requested a contract file download
+    $download = CRM_Utils_Request::retrieve('ct_dl', 'String', CRM_Core_DAO::$_nullObject, FALSE, '', 'GET');
+    if (!empty($download)) {
+      CRM_Contract_Utils::downloadContractFile($download);
```

Comment:
Ok, I've had another look at this.  CRM_Utils_System::download doesn't really do much to help - you still have to put the file into a buffer, set the mime type etc.  With my function, it does all that for you, and loads the file to a buffer in 2KB chunks, reports the total file size to the client browser via a header and downloads the file.  By the way, it's the same function I used in com.cividesk.normalize for the log files.  I use Chrome and Firefox.